crates/sui-storage/src/key_value_store.rs

```diff
@@ -0,0 +1,171 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Immutable key/value store trait for storing/retrieving transactions, effects, and events
+//! to/from a scalable.
+
+use async_trait::async_trait;
+use sui_types::digests::{TransactionDigest, TransactionEffectsDigest, TransactionEventsDigest};
+use sui_types::effects::{TransactionEffects, TransactionEvents};
+use sui_types::error::SuiResult;
+use sui_types::transaction::Transaction;
+use tracing::error;
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum Key {
+    Tx(TransactionDigest),
+    Fx(TransactionEffectsDigest),
+    Events(TransactionEventsDigest),
+}
+
+impl From<TransactionDigest> for Key {
+    fn from(tx: TransactionDigest) -> Self {
+        Key::Tx(tx)
+    }
+}
+
+impl From<TransactionEffectsDigest> for Key {
+    fn from(fx: TransactionEffectsDigest) -> Self {
+        Key::Fx(fx)
+    }
+}
+
+impl From<TransactionEventsDigest> for Key {
+    fn from(events: TransactionEventsDigest) -> Self {
+        Key::Events(events)
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum Value {
+    Tx(Box<Transaction>),
+    Fx(Box<TransactionEffects>),
```

Comment:
do we want to keep it separate or store it as a bundle with Transaction itself?

Reply:
you mean something like

     struct Value(Transaction, TransactionEffects)

My inclination is to keep it separate, there are definitely cases that only read one of the values

Reply:
Yes